### PR TITLE
improve hadoop_open docs

### LINKED ISSUE #4643: give an example of using hadoop_open
`hadoop_open` needs an example of writing a pandas data frame directly to a google bucket

https://hail.is/docs/stable/utils/index.html?highlight=hadoop_copy#hail.utils.hadoop_open

### PR DESCRIPTION
**Pull request by danking**
Requested by Andrea

resolves #4643

### CODE REVIEW

File: hail/python/hail/utils/hadoop_utils.py [hunk 1]
```diff
@@ -84,10 +95,15 @@ def hadoop_copy(src, dest):
     """Copy a file through the Hadoop filesystem API.
     Supports distributed file systems like hdfs, gs, and s3.
 
+    .. Hint:: Try using ``hadoop_open`` first, it's more efficient!
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/python/hail/utils/hadoop_utils.py]: I also don't think hadoop open is more efficient. It totally depends what you're doing. If you're writing a lot of data, hadoop_open will be way *less* efficient

File: hail/python/hail/utils/hadoop_utils.py [hunk 2]
```diff
@@ -17,14 +17,25 @@ def hadoop_open(path: str, mode: str = 'r', buffer_size: int = 8192):
...
     ...     for line in f:
     ...         print(line.strip())
 
+    Write two lines directly to a file in Google Cloud Storage:
+
     >>> with hadoop_open('gs://my-bucket/notes.txt', 'w') as f: # doctest: +SKIP
     ...     f.write('result1: %s\\n' % result1)
     ...     f.write('result2: %s\\n' % result2)
 
+    Unpack a packed Python struct directly from a file in Google Cloud Storage:
```
[the remark this answers is not in the document]
Comment (danking) [reply hail/python/hail/utils/hadoop_utils.py]: Nope.

File: hail/python/hail/utils/hadoop_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danking) [reply hail/python/hail/utils/hadoop_utils.py]: It's less efficient to write the CSV to disk then copy it up to GS?

**Comment by danking** [comment]
@tpoterba I don't quite see why anyone should use `hadoop_copy` if their tool supports exporting to a file handle. Unless they're doing some random access, but I think that's much rarer than `df.to_csv(f)` and using `hadoop_copy` for that is silly (AFAIK?).

**Comment by tpoterba** [comment]
our hadoop_open file handles are really slow. ~10M/s

**Comment by danking** [comment]
!? wtf? do we have any idea why?

**Comment by tpoterba** [comment]
yeah, look at the implementation - py4j

**Comment by danking** [comment]
Well. Should anyone ever use hadoop_open then?

**Comment by tpoterba** [comment]
yes, for small files it's easier

**Comment by danking** [comment]
@tpoterba if you're OK with this, I think @jigold is also OK, so could you approve :3

**Comment by danking** [comment]
@jigold I had to switch back to `.. Caution::` from the section header because Sphinx does not like section headers called "Caution". I'm attempting to get the [admonition effect](https://sphinx-rtd-theme.readthedocs.io/en/latest/demo/demo.html#admonitions).

**Comment by tpoterba** [comment]
You can use `Warning`

**Comment by tpoterba** [comment]
that's what we do elsewhere. It's a synonym and will render as an admonition

**Comment by danking** [comment]
sphinx never ceases to confuse me.

File: hail/python/hail/utils/hadoop_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/python/hail/utils/hadoop_utils.py]: I think it came from BlockMatrix stuff?

File: hail/python/hail/utils/hadoop_utils.py [hunk 3]
```diff
@@ -84,10 +101,22 @@ def hadoop_copy(src, dest):
     """Copy a file through the Hadoop filesystem API.
     Supports distributed file systems like hdfs, gs, and s3.
 
+    Hint
```
Comment (tpoterba) [inline hail/python/hail/utils/hadoop_utils.py]: hint is an invalid directive - change to Note? and move below the examples?

Comment (danking) [reply hail/python/hail/utils/hadoop_utils.py]: What am I missing, why don't all admonitions listed by Read The Docs work? https://sphinx-rtd-theme.readthedocs.io/en/latest/demo/demo.html#admonitions

**Pull request closed without merging**